[AMDGPU] Split struct kernel arguments

yxsamliu opened the pull request:
Integrate split kernel arguments logic into the preload pass. When a struct kernel argument has fields that can be individually preloaded, split the argument into separate scalar/vector arguments for more efficient SGPR preloading.

Key features:
- Layout-preserving mode (default): Fields stay at original offsets, uses backup declaration to preserve metadata for CLR compatibility
- Layout-changing mode (-amdgpu-kernarg-layout-change): Reorders fields for optimal packing
- Only splits when target supports preload and args fit in SGPRs
- Handles nested structs, arrays with constant indices, vectors
- Variable-index arrays and address-taken args are not split

The backup declaration approach allows stock CLR to work without patching by emitting original struct metadata.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHSAMetadataStreamer.cpp:
```diff
@@ -357,17 +357,51 @@ void MetadataStreamerMsgPackV4::emitKernelArg(const Argument &Arg,
+  // unless we find it in function attribute amdgpu-argument-mapping.
+  unsigned OriginalArgIndex = ~0U;
+  uint64_t OriginalArgOffset = 0;
+  if (Func->hasFnAttribute("amdgpu-argument-mapping")) {
```
Just do the getFnAttribute and test if it's valid

yxsamliu replied:
will do

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+
+bool AMDGPUSplitKernelArguments::processFunction(Function &F) {
+  const DataLayout &DL = F.getParent()->getDataLayout();
+  LLVM_DEBUG(dbgs() << "Entering AMDGPUSplitKernelArguments::processFunction "
+                    << F.getName() << '\n');
```
Remove this, this should be the pass manager's job

yxsamliu replied:
will remove

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  LLVM_DEBUG(dbgs() << "Entering AMDGPUSplitKernelArguments::processFunction "
+                    << F.getName() << '\n');
+  if (F.isDeclaration()) {
+    LLVM_DEBUG(dbgs() << "Function is a declaration, skipping\n");
```
```suggestion
```
Only noise

yxsamliu replied:
will remove

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+
+    Type *ArgTy = Arg.getParamByRefType();
+    StructType *ST = dyn_cast<StructType>(ArgTy);
+    if (!ST) {
+      NewArgTypes.push_back(Arg.getType());
+      LLVM_DEBUG(dbgs() << "not a struct\n");
+      // Include mapping if indices have changed
+      if (NewArgIndex != OriginalArgIndex)
+        NewArgMappings.push_back(
+            std::make_tuple(NewArgIndex, OriginalArgIndex, 0));
+      ++NewArgIndex;
+      ++OriginalArgIndex;
+      continue;
+    }
```
Why do you need to handle struct arguments? We should stop using these completely, and only use byref. I don't want to add additional code that's processing direct struct uses

yxsamliu replied:
This code is not for direct struct-type argument. When the control flow reaches here, we are looking at the pointee type of a byref argument. A byref argument is a pointer type in LLVM IR. It can point to struct type or non-struct type. This pass only optimizes byref struct type.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+
+  // Add the mapping of the new arguments to the old arguments as a function
+  // attribute in the format "NewArgIndex:OriginalArgIndex:Offset,..."
+  std::string MappingStr;
```
raw_svector_ostream to build string

yxsamliu replied:
will do

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+      if (GEP->use_empty()) {
+        GEP->eraseFromParent();
+      } else {
+        GEP->replaceAllUsesWith(UndefValue::get(GEP->getType()));
```
Use poison

yxsamliu replied:
done

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  for (Argument &Arg : F.args()) {
+    if (ArgToLoadsMap.count(&Arg)) {
+      for (LoadInst *LI : ArgToLoadsMap[&Arg]) {
+        std::string OldName = LI->getName().str();
+        LI->setName(OldName + ".old");
+        NewArgIt->setName(OldName);
```
takeName

yxsamliu replied:
will do

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHSAMetadataStreamer.cpp:
```diff
@@ -409,6 +443,12 @@ void MetadataStreamerMsgPackV4::emitKernelArg(
       Arg[".is_pipe"] = Arg.getDocument()->getNode(true);
   }
 
+  // Add original argument index and offset to the metadata
+  if (OriginalArgIndex != ~0U) {
+    Arg[".original_arg_index"] = Arg.getDocument()->getNode(OriginalArgIndex);
+    Arg[".original_arg_offset"] = Arg.getDocument()->getNode(OriginalArgOffset);
+  }
```
I don't want to give meaning to the index in the IR signature. It already doesn't have a 1:1 mapping with the source language. The ABI should only be in terms of sizes, alignments and offsets

yxsamliu replied:
> I don't want to give meaning to the index in the IR signature. It already doesn't have a 1:1 mapping with the source language. The ABI should only be in terms of sizes, alignments and offsets

In clang codegen, clang may make changes to the kernel signature so that it is no longer the same as in the source code. However, the kernel arg index does play critical role in the contract between the compiler and runtime. Basically, the kernel launching API accepts an array of pointers pointing to each kernel argument, and it lays out each kernel argument one by one into the kernel argument segment based on HSA metadata. This is because in host compilation it is not always possible to know the exact size, align, and offset of each kernel argument that is determined in device compilation, therefore those information are passed to the runtime through HSA metadata, and used during kernel launching.

Due to this, it is critical that mid-end and backend cannot change the number of kernel arguments without have a way to tell the runtime of such changes. Before this PR, this was maintained because a kernel has external linkage with default visibility in LLVM IR, therefore IPO cannot change their signatures. In this PR, we want to optimize the arguments passing from runtime to firmware by omitting the useless arguments and useless fields in struct-type arguments, since the runtime sees an array of pointers to arguments, we have to tell runtime which portion of which argument to keep. This function attribute is intended for generating HSA metadata for the runtime.

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-attributes --check-globals all --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-split-kernel-arguments -amdgpu-enable-split-kernel-args -verify-machineinstrs < %s | FileCheck %s
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-split-kernel-arguments -amdgpu-enable-split-kernel-args -verify-machineinstrs < %s | llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 -verify-machineinstrs | FileCheck --check-prefix=GCN %s
```
Remove all the -verify-machineinstrs, especially for opt where it doesn't do anything. Also use temporary file if you want to codegen the pass output

yxsamliu replied:
will do

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+  %coerce1.sroa.217.0.copyload = load i8, ptr addrspace(4) %coerce1.sroa.217.0..sroa_idx, align 4
+  %coerce1.sroa.318.0..sroa_idx = getelementptr inbounds nuw i8, ptr addrspace(4) %1, i64 56
+  %coerce1.sroa.318.0.copyload = load ptr, ptr addrspace(4) %coerce1.sroa.318.0..sroa_idx, align 8
+  %2 = addrspacecast ptr %coerce1.sroa.318.0.copyload to ptr addrspace(1)
```
Use named values

yxsamliu replied:
will do

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-attributes --check-globals all --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-split-kernel-arguments -amdgpu-enable-split-kernel-args -verify-machineinstrs < %s | FileCheck %s
```
Missing checks of the IR pass output

yxsamliu replied:
Can you elaborate? I have FileCheck for opt output and autogenerated check lines.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+using namespace llvm;
+
+namespace {
+static llvm::cl::opt<bool> EnableSplitKernelArgs(
+    "amdgpu-enable-split-kernel-args",
+    llvm::cl::desc("Enable splitting of AMDGPU kernel arguments"),
+    llvm::cl::init(false));
```
Don't need the llvm::s

yxsamliu replied:
will do

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// \file This pass flats struct-type kernel arguments. It eliminates unused
+// fields and only keeps used fields. The objective is to facilitate preloading
+// of kernel arguments by later passes.
```
Is it really worth having an ABI breaking pass for this? Why can't the use just reorder the arguments? 

This needs substantial unit test coverage, preferably in the llvm-test-suite 

I also think this is leaning too heavily on the runtime dispatch inherited from OpenCL, really we ought to be able to emit directly callable signatures without going through the runtime metadata to set up the dispatch arguments

yxsamliu replied:
add a test to llvm-test-suite: https://github.com/llvm/llvm-test-suite/pull/230

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+
+  CallingConv::ID CC = F.getCallingConv();
+  if (CC != CallingConv::AMDGPU_KERNEL || F.arg_empty()) {
+    LLVM_DEBUG(dbgs() << "non-kernel or arg_empty\n");
```
```suggestion
```

Noise

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  SmallVector<Argument *, 8> StructArgs;
+  SmallVector<Type *, 8> NewArgTypes;
+
+  auto convertAddressSpace = [](Type *Ty) -> Type * {
+    if (auto *PtrTy = dyn_cast<PointerType>(Ty)) {
+      if (PtrTy->getAddressSpace() == AMDGPUAS::FLAT_ADDRESS) {
+        return PointerType::get(PtrTy->getContext(), AMDGPUAS::GLOBAL_ADDRESS);
+      }
+    }
+    return Ty;
+  };
```
Side hack, probably shouldn't do this here

yxsamliu replied:
This is to keep the addr space changed by AMDGPUPromoteKernelArgumentsPass. For pointer type fields in a byref struct argument, AMDGPUPromoteKernelArgumentsPass inserts an addr space cast since global load/store is faster than flat load/store: https://godbolt.org/z/vErjhW7f9

When we flatten the byref struct argument, the pointer-type field becomes a new argument, we want to make it a global pointer type argument so that we can eliminate the redundant addr space cast to make the IR cleaner.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  // Add the mapping of the new arguments to the old arguments as a function
+  // attribute in the format "NewArgIndex:OriginalArgIndex:Offset,..."
+  std::string MappingStr;
+  for (const auto &Info : NewArgMappings) {
```
This format is going to be really hard to follow. If you really want to mark up an argument, it should be directly attached to the argument

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+      if (!AllLoadsOrGEPs)
+        break;
+    }
+    LLVM_DEBUG(dbgs() << "  AllLoadsOrGEPs: "
+                      << (AllLoadsOrGEPs ? "true" : "false") << "\n");
```
memcpy also relevant

yxsamliu replied:
memcpy accepts a pointer to a region in the kernel argument segment. Usually it is not pointing to a field of a primitive type that can be preloaded. As such, my pass will not optimize the byref struct-type argument if itself or its field is accessed by memcpy.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+    PointerType *PT = dyn_cast<PointerType>(Arg.getType());
+    if (!PT) {
+      NewArgTypes.push_back(Arg.getType());
+      LLVM_DEBUG(dbgs() << "not a pointer\n");
```
```suggestion
```

yxsamliu replied:
will do

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+          std::make_tuple(NewArgIndex, OriginalArgIndex, 0));
+      ++NewArgIndex;
+      ++OriginalArgIndex;
+      LLVM_DEBUG(dbgs() << "use empty\n");
```
```suggestion
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  unsigned OriginalArgIndex = 0;
+  unsigned NewArgIndex = 0;
+  for (Argument &Arg : F.args()) {
+    LLVM_DEBUG(dbgs() << "Processing argument: " << Arg << "\n");
```
```suggestion
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+    const bool IsByRef = Arg.hasByRefAttr();
+    if (!IsByRef) {
+      NewArgTypes.push_back(Arg.getType());
+      LLVM_DEBUG(dbgs() << "not byref\n");
```
```suggestion
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+      continue;
+    }
+
+    const bool IsByRef = Arg.hasByRefAttr();
```
Pass should exclusively handle byref arguments if anything

yxsamliu replied:
Yes it is. Non-byref argument is skipped in this loop

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  // Erase GEPs
+  for (auto &Entry : ArgToGEPsMap) {
+    for (GetElementPtrInst *GEP : Entry.second) {
+      if (GEP->use_empty()) {
```
Don't need to treat this as a special case

yxsamliu replied:
will remove

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,373 @@
+  }
+
+  // Replace LoadInsts with new arguments
+  for (auto &Entry : ArgToLoadsMap) {
+    for (LoadInst *LI : Entry.second) {
+      Value *NewArg = VMap[LI];
+      LI->replaceAllUsesWith(NewArg);
+      LI->eraseFromParent();
+    }
+  }
+
+  // Erase GEPs
+  for (auto &Entry : ArgToGEPsMap) {
```
Shouldn't need to consider different user types separately

arsenm reviewed (changes requested):
I'd rather avoid having an ABI breaking pass, and I definitely want to avoid spreading the handling of non-byref kernel arguments. This has overlap with the existing IR expansion of non-byref arguments

yxsamliu commented:
> I'd rather avoid having an ABI breaking pass, and I definitely want to avoid spreading the handling of non-byref kernel arguments. This has overlap with the existing IR expansion of non-byref arguments

For HIP programs that launches the kernel, there is no ABI change since the arguments passed to the kernel launching triple chevron do not change. This is an internal optimization of kernel launching procedure by skipping unused fields in struct type kernel arguments so that only used fields are preloaded. In a sense, it is similar to the arguments promotion IPO, but happening between the compiler and the runtime through the newly added kernel argument metadata. Basically, it tells runtime which chunk of kernel arguments are to be kept and then preloaded. Since the optimization only happens late in the LLVM pipeline, it won't proliferate non-byref kernel arguments in FE or middle end.

shiltian commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+; CHECK-NEXT:    ret void
+;
+entry:
+  %coerce.sroa.1.0..sroa_idx = getelementptr inbounds nuw i8, ptr addrspace(4) %0, i64 8
```
can you also rename those variables to make the test case easier to understand?

yxsamliu replied:
will do

shiltian commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+}
+
+;.
+; CHECK: attributes #[[ATTR0]] = { "amdgpu-argument-mapping"="2:2:8,4:4:56,5:4:44,6:4:12" }
```
Can you add the document of `amdgpu-argument-mapping` as well?

yxsamliu replied:
will do

shiltian commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,120 @@
+;  b.a2.p[2] = a.l + b.a2.c;
+;}
+;
+%struct.A = type { i32, i8, i64, ptr }
```
Do we (compiler and/or runtime) have special handling of 64-bit byval? In OpenMP we pass all arguments as `void *`, which is 64-bit so if anything smaller than that, it can be passed as byval and then we do a type punning in the code. Do we have something similar in HIP?

yxsamliu replied:
https://godbolt.org/z/3aGMEjvMK

If a struct size is not greater than 64 bit and has only one field, and in HIP source code it is passed by value, clang will change it to a scalar parameter in the kernel signature. In this case, it is a normal scalar argument in LLVM IR and this pass does not handle it.

Otherwise, clang use byref ptr to struct type for it in the kernel signature, and my pass will handle it.

yxsamliu commented:
gentle ping

yxsamliu commented:
ping

shiltian commented:
> it is similar to the arguments promotion IPO

which we (at least @arsenm and I) want to get rid of and want to do it right in the front end instead. :-D

However, IMO this pass is fine. We are not doing something "less optimal" in the first place and then try to correct it later, unlike the kernel argument promotion. It's hard to do this kind of reasoning in the front end.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+
+  CallingConv::ID CC = F.getCallingConv();
+  if (CC != CallingConv::AMDGPU_KERNEL || F.arg_empty())
+    return false;
```
this check can be done when we iterate all functions in the module

yxsamliu replied:
will move it there

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+  SmallVector<Argument *, 8> StructArgs;
+  SmallVector<Type *, 8> NewArgTypes;
+
+  auto convertAddressSpace = [](Type *Ty) -> Type * {
```
This handling is similar to kernel argument promotion but takes a more aggressive approach. I don't think we should scatter this functionality across multiple places.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+      continue;
+    }
+
+    const bool IsByRef = Arg.hasByRefAttr();
+    if (!IsByRef) {
```
```suggestion
    if (!Arg.hasByRefAttr()) {
```

yxsamliu replied:
will do

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    Type *ArgTy = Arg.getParamByRefType();
+    StructType *ST = dyn_cast<StructType>(ArgTy);
+    if (!ST) {
```
```suggestion
    if (!isa<StructType>(ArgTy)) {
```

`ST` doesn't seem to be used in the following.

shiltian reviewed (commented):
Another question is, does it affect explicit kernel launch via `hipLaunchKernelGGL` or even the lower level use via HSA?

yxsamliu commented:
> Another question is, does it affect explicit kernel launch via `hipLaunchKernelGGL` or even the lower level use via HSA?

No. The kernel signature does not change from the user pointer of view. It only needs HIP runtime change about laying out the kernel arg segment. It does not need HSA or firmware change.

yxsamliu commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+
+bool AMDGPUSplitKernelArguments::processFunction(Function &F) {
+  const DataLayout &DL = F.getParent()->getDataLayout();
+  if (F.isDeclaration())
```
will remove

yxsamliu commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,349 @@
+    FunctionsToProcess.push_back(&F);
+  }
+
+  for (Function *F : FunctionsToProcess) {
+    if (F->isDeclaration())
+      continue;
```
no. will remove that

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+    "amdgpu-enable-split-kernel-args",
+    cl::desc("Enable splitting of AMDGPU kernel arguments"), cl::init(false));
+
+class AMDGPUSplitKernelArguments : public ModulePass {
```
IIUC this is an implementation class. In that case, I don't think it needs to inherit `ModulePass`.

yxsamliu replied:
This is for supporting the legacy pass manager. So far llc still use the legacy pass manager by default and AMDGPU backend supports it. I am not sure whether I only need to add this pass to the new pass manager, therefore I added this pass to both the legacy and the new pass manager.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+
+    if (!isa<StructType>(Arg.getParamByRefType())) {
+      NewArgTypes.push_back(Arg.getType());
+      // Include mapping if indices have changed
+      if (NewArgIndex != OriginalArgIndex)
+        NewArgMappings.push_back(
+            std::make_tuple(NewArgIndex, OriginalArgIndex, 0));
```
nit: make it a lambda since they are just exactly the same code for three times.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+          LLVM_DEBUG(dbgs() << "    GEP User: " << *GEPUser << "\n");
+          if (auto *GEPLoad = dyn_cast<LoadInst>(GEPUser)) {
+            Loads.push_back(GEPLoad);
+          } else {
```
How about a nested GEP?

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+        uint64_t Offset = 0;
+        if (auto *GEP = dyn_cast<GetElementPtrInst>(LI->getPointerOperand())) {
+          APInt OffsetAPInt(DL.getPointerSizeInBits(), 0);
+          if (GEP->accumulateConstantOffset(DL, OffsetAPInt))
```
I think there is some issue here. What if offset is not a constant? Need a test case to cover this as well.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+  NewArgIndex = 0;
+  for (Argument &Arg : F.args()) {
+    if (ArgToLoadsMap.count(&Arg)) {
+      for (LoadInst *LI : ArgToLoadsMap[&Arg]) {
+        (void)LI;
```
```suggestion
      for ([[maybe_used]] LoadInst *LI : ArgToLoadsMap[&Arg]) {
```

yxsamliu replied:
will do

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,328 @@
+  DenseMap<Value *, Value *> VMap;
+  auto NewArgIt = NewF->arg_begin();
+  for (Argument &Arg : F.args()) {
+    if (ArgToLoadsMap.count(&Arg)) {
```
```suggestion
    if (ArgToLoadsMap.contains(&Arg)) {
```

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUHSAMetadataStreamer.cpp:
```diff
@@ -409,6 +430,12 @@ void MetadataStreamerMsgPackV4::emitKernelArg(
 
+  // Add original argument index and offset to the metadata
+  if (OriginalArgIndex != ~0U) {
+    Arg[".original_arg_index"] = Arg.getDocument()->getNode(OriginalArgIndex);
```
Honestly, I agree with @arsenm that this is indeed an ABI breaking change, even though the ABI change seems to be transparent to end users.

shiltian commented on llvm/test/CodeGen/AMDGPU/amdgpu-split-kernel-args.ll:
```diff
@@ -0,0 +1,144 @@
+; CHECK-NEXT:    store i32 [[CONV15]], ptr addrspace(1) [[ARRAYIDX]], align 4
+; CHECK-NEXT:    ret void
+;
+  ptr addrspace(1) noundef writeonly captures(none) initializes((0, 4)) %out,
```
we need test case for all checks in the `for` loop to make sure they work properly.

yxsamliu replied:
will do

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -903,6 +904,7 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
     if (Level != OptimizationLevel::O0) {
       if (!isLTOPreLink(Phase)) {
         AMDGPUAttributorOptions Opts;
+        MPM.addPass(AMDGPUSplitKernelArgumentsPass());
         MPM.addPass(AMDGPUAttributorPass(*this, Opts, Phase));
```
nit:

```suggestion
        MPM.addPass(AMDGPUAttributorPass(*this, Opts, Phase));
        MPM.addPass(AMDGPUSplitKernelArgumentsPass());
```

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1293,6 +1296,9 @@ void AMDGPUPassConfig::addIRPasses() {
     addPass(createAMDGPULowerModuleLDSLegacyPass(&TM));
   }
 
+  if (TM.getOptLevel() > CodeGenOptLevel::None) {
+    addPass(createAMDGPUSplitKernelArgumentsPass());
+  }
```
```suggestion
  if (TM.getOptLevel() > CodeGenOptLevel::None)
    addPass(createAMDGPUSplitKernelArgumentsPass());
```

yxsamliu replied:
will do

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp:
```diff
@@ -0,0 +1,334 @@
+}
+
+PreservedAnalyses
+AMDGPUSplitKernelArgumentsPass::run(Module &M, ModuleAnalysisManager &AM) {
```
Is this pass safe to be run multiple times?

shiltian reviewed (commented):
The pass generally looks good to me. Let's wait for @arsenm regarding the ABI break concern.

shiltian commented:
> > Another question is, does it affect explicit kernel launch via `hipLaunchKernelGGL` or even the lower level use via HSA?
> 
> No. The kernel signature does not change from the user pointer of view. It only needs HIP runtime change about laying out the kernel arg segment. It does not need HSA or firmware change.

I mean, if users launch a kernel using HSA runtime instead of HIP (which is what OpenMP is doing), is it gonna break?

yxsamliu commented:
> > > Another question is, does it affect explicit kernel launch via `hipLaunchKernelGGL` or even the lower level use via HSA?
> > 
> > 
> > No. The kernel signature does not change from the user pointer of view. It only needs HIP runtime change about laying out the kernel arg segment. It does not need HSA or firmware change.
> 
> I mean, if users launch a kernel using HSA runtime instead of HIP (which is what OpenMP is doing), is it gonna break?

OpenMP does not seem to use explicit kernel args although it uses implicit kernel args https://github.com/llvm/llvm-project/blob/main/offload/plugins-nextgen/amdgpu/src/rtl.cpp#L3393 . Since it does not use explicit kernel args, it is not affected by this patch. Even if it decides to use explicit kernel args, based on the libomptarget code, it needs to layout kernel arguments from host into kernel argument segment as dictated by code object metadata, before passing the pointer to the kernel arg segment to HSA runtime, in a similar way as HIP runtime does. Therefore, as long as it correctly follow the code object metadata, it will be able to launch the kernel transformed by this pass correctly with HSA runtime.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -935,6 +937,7 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
             PM.addPass(InternalizePass(mustPreserveGV));
             PM.addPass(GlobalDCEPass());
           }
+          PM.addPass(AMDGPUSplitKernelArgumentsPass());
```
Also, I think we might want to think carefully where to add the pass. Will it potentially benefit other passes? It looks like for now it is added to where the attributor pass runs, which is actually almost the end of the pipeline. Is it expected?

yxsamliu commented:
@arsenm Any further concerns or comments about this PR? Thanks.

arsenm reviewed (commented):
I think this should be integrated directly into the preloading, not as a separate step. The new inline kernel arguments should be introduced if and only if they will be preloaded 

This should also not try to sort or reorder arguments for optimal packing without an option (and preferably that part would be a separate patch)

yxsamliu committed:
[AMDGPU] Split struct kernel arguments for preloading

Integrate split kernel arguments logic into the preload pass. When a
struct kernel argument has fields that can be individually preloaded,
split the argument into separate scalar/vector arguments for more
efficient SGPR preloading.

Key features:
- Layout-preserving mode (default): Fields stay at original offsets,
  uses backup declaration to preserve metadata for CLR compatibility
- Layout-changing mode (-amdgpu-kernarg-layout-change): Reorders fields
  for optimal packing
- Only splits when target supports preload and args fit in SGPRs
- Handles nested structs, arrays with constant indices, vectors
- Variable-index arrays and address-taken args are not split

The backup declaration approach allows stock CLR to work without
patching by emitting original struct metadata.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUPreloadKernelArguments.cpp:
```suggestion
      NewArgMappings.emplace_back(NewArgIndex, OriginalArgIndex, /*Offset=*/0);
```
